### PR TITLE
network: fold ipv4ll fallback modes into normal ipv4ll addressing

### LINKED ISSUE #13316: Networkd: LinkLocalAddressing=fallback with MaxAttempts=infinity
I came across a requirement why the link-local address exists with other IP addresses for IPv4.

Suppose if I enable the link-local address through systemd-networkd then as per the existing

behavior, link-local address(169.254.*.*) will always exist on the interface irrespective of whether static/dynamic address exists or not.

The expectation is if either static/DHCP address exist then the zero-config address should not come even if the zeroconf functionality is enabled.

Is it correct expectation? is it aligned with other operating systems which is not using systemd-networkd as the network manager?

### PR DESCRIPTION
**Pull request by yuwata**
Replaces #17290.
Fixes #13316.

### CODE REVIEW

**Comment by yuwata** [comment]
Yey, now all CentOS CIs are green.

File: src/network/networkd-dhcp4.c [hunk 1]
```diff
@@ -1105,20 +1105,37 @@ static int dhcp4_handler(sd_dhcp_client *client, int event, void *userdata) {
                                 return r;
                         }
 
+                        if (link->ipv4ll) {
+                                r = sd_ipv4ll_stop(link->ipv4ll);
```
Comment (keszybz) [inline src/network/networkd-dhcp4.c]: `sd_ipv4ll_stop` immediately returns if the argument is NULL, so this `if` check in unnecessary.

A question though: does this result in the address being immediately dropped or just the address not being preferred anymore? If we drop the address immediately, existing connections will be broken.

**Review by keszybz — COMMENTED** [review]
Thanks for picking this up.

**Comment by yuwata** [comment]
@keszybz Thank you for the comment. Now, the IPv4LL address is dropped after the DHCPv4 address becomes ready. PTAL.

**Comment by keszybz** [comment]
So... what is expected behaviour after we finally acquire the lease? I get the following:
```console
$ while ip a show dev host0; do sleep 1; done
2: host0@if40: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether 96:aa:3a:7b:d4:b9 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 169.254.40.164/16 brd 169.254.255.255 scope link host0
       valid_lft forever preferred_lft forever
    inet6 fe80::94aa:3aff:fe7b:d4b9/64 scope link 
       valid_lft forever preferred_lft forever
...
2: host0@if40: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether 96:aa:3a:7b:d4:b9 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 169.254.40.164/16 brd 169.254.255.255 scope link host0
       valid_lft forever preferred_lft forever
    inet 192.168.102.53/28 brd 192.168.102.63 scope global dynamic host0
       valid_lft 3600sec preferred_lft 3600sec
    inet6 fe80::94aa:3aff:fe7b:d4b9/64 scope link 
       valid_lft forever preferred_lft forever
...
2: host0@if40: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether 96:aa:3a:7b:d4:b9 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 169.254.40.164/16 brd 169.254.255.255 scope link host0
       valid_lft forever preferred_lft forever
    inet 192.168.102.53/28 brd 192.168.102.63 scope global dynamic host0
       valid_lft 3582sec preferred_lft 3582sec
    inet6 fe80::94aa:3aff:fe7b:d4b9/64 scope link 
       valid_lft forever preferred_lft forever
```
I.e. the LL address will remain forever. Shouldn't we mark it non-preferred and drop after a minute or so?

**Comment by keszybz** [comment]
Anyway, let's merge this. We can improve on the lifetime later.

>  Shouldn't we mark it non-preferred and drop after a minute or so?

If the kernel allow that, I'd like to go from
```
inet 169.254.40.164/16 brd 169.254.255.255 scope link host0
       valid_lft forever preferred_lft forever
```
to
```
inet 169.254.40.164/16 brd 169.254.255.255 scope link host0
       valid_lft 180 preferred_lft 0
```
when the DHCPv4 address is acquired.

**Comment by yuwata** [comment]
The IPv4LL address must be removed. I will prepare a follow-up PR.

**Comment by BillPlunkett** [comment]
I've been following this discussion for a while, but not 100% sure I understand where things stand.

The desired behavior (for me) is:

network configured in DHCP client mode
DHCP server not present or unresponsive, so ipv4ll address assigned
DHCP server (re)appears, so ipv4ll address removed and DHCP address assigned
From above, it seems like this behavior has been implemented in v247 with the exception that the ipv4ll address will remain after a DHCP address is assigned in step 3. Assuming that is true, what are the ramifications?

**Comment by keszybz** [comment]
The expected behaviour is how you describe it.  #17798 even included a test that the ipv4ll address is dropped. Sans bugs, we should be good.

**Comment by ratagupt** [comment]
@yuwata : Can you explian the behaviour here? Do we still have the fall back option or it has been deprecated.

**Comment by yuwata** [comment]
@ratagupt The fall back option is now deprecated. IPv4LL is assigned only when DHCP client is disabled or the client does not receive any response from servers.

**Comment by ratagupt** [comment]
@yuwata : I tried it on systemd v247 and it didn't work with linklocaladdressing=yes, I can see that LL address was always there even after getting the DHCP address.

**Comment by yuwata** [comment]
Could you open a new issue page with debugging logs? Otherwise, we can easily forget your comment in this already merged and closed page.

**Comment by BillPlunkett** [comment]
I was thinking the ipv4 dhcp/linklocal fallback feature made it into systemd v247, but it seems not.  I grabbed yocto-hardknott (latest yocto release), which uses systemd v247.6 and it's pretty clear the changes are not there.  From https://layers.openembedded.org/layerindex/recipe/5979/ it looks as if v247 is the most recent version available in any yocto release.  I am tempted to patch the 247.6 networkd-dhcp4.c to get this feature.  Any advice or suggestions?

**Comment by nikhilsunilkumar** [comment]
> I was thinking the ipv4 dhcp/linklocal fallback feature made it into systemd v247, but it seems not. I grabbed yocto-hardknott (latest yocto release), which uses systemd v247.6 and it's pretty clear the changes are not there. From https://layers.openembedded.org/layerindex/recipe/5979/ it looks as if v247 is the most recent version available in any yocto release. I am tempted to patch the 247.6 networkd-dhcp4.c to get this feature. Any advice or suggestions?

Any updates on this ?

**Pull request merged**